File: android-app/app/src/main/java/com/plantgotchi/app/model/Course.kt
package com.plantgotchi.app.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "courses",
    indices = [Index(value = ["slug"], unique = true), Index(value = ["creator_id"])]
)
data class Course(
    @PrimaryKey
    val id: String,

    @ColumnInfo(name = "creator_id")
    val creatorId: String,

    @ColumnInfo(name = "creator_name")
    val creatorName: String? = null,

    val title: String,

    val slug: String,

    val description: String? = null,

    @ColumnInfo(name = "cover_image_url")
    val coverImageUrl: String? = null,

    @ColumnInfo(name = "price_cents")
    val priceCents: Int = 0,

    val currency: String = "USD",

    val status: String = "published",

    @ColumnInfo(name = "enrollment_count")
    val enrollmentCount: Int? = null,

    @ColumnInfo(name = "created_at")
    val createdAt: String? = null,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String? = null,
)

@Entity(
    tableName = "course_phases",
    foreignKeys = [ForeignKey(entity = Course::class, parentColumns = ["id"], childColumns = ["course_id"], onDelete = ForeignKey.CASCADE)],
    indices = [Index(value = ["course_id", "sort_order"])]
)
data class CoursePhase(
    @PrimaryKey
    val id: String,

    @ColumnInfo(name = "course_id")
    val courseId: String,

    val title: String,

    val description: String? = null,

    @ColumnInfo(name = "sort_order")
    val sortOrder: Int,

    @ColumnInfo(name = "created_at")
    val createdAt: String? = null,
)

@Entity(
    tableName = "phase_modules",
    foreignKeys = [ForeignKey(entity = CoursePhase::class, parentColumns = ["id"], childColumns = ["phase_id"], onDelete = ForeignKey.CASCADE)],
    indices = [Index(value = ["phase_id", "sort_order"])]
)
data class PhaseModule(
    @PrimaryKey
    val id: String,

    @ColumnInfo(name = "phase_id")
    val phaseId: String,

    val title: String,

    val description: String? = null,

    @ColumnInfo(name = "sort_order")
    val sortOrder: Int,

    @ColumnInfo(name = "created_at")
    val createdAt: String? = null,
)

@Entity(
    tableName = "module_content_blocks",
    foreignKeys = [ForeignKey(entity = PhaseModule::class, parentColumns = ["id"], childColumns = ["module_id"], onDelete = ForeignKey.CASCADE)],
    indices = [Index(value = ["module_id", "sort_order"])]
)
data class ContentBlock(
    @PrimaryKey
    val id: String,

    @ColumnInfo(name = "module_id")
    val moduleId: String,

    @ColumnInfo(name = "block_type")
    val blockType: String, // "video", "text", "quiz"

    @ColumnInfo(name = "sort_order")
    val sortOrder: Int,

    val content: String, // JSON string

    @ColumnInfo(name = "created_at")
    val createdAt: String? = null,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String? = null,
)

File: ios-app/Plantgotchi/Models/Course.swift
import Foundation
import GRDB

// MARK: - Course

struct Course: Codable, Identifiable, Equatable {
    var id: String
    var creatorId: String
    var creatorName: String?
    var title: String
    var slug: String
    var description: String?
    var coverImageUrl: String?
    var priceCents: Int
    var currency: String
    var status: String
    var enrollmentCount: Int?
    var createdAt: String?
    var updatedAt: String?

    init(
        id: String = UUID().uuidString,
        creatorId: String,
        creatorName: String? = nil,
        title: String,
        slug: String,
        description: String? = nil,
        coverImageUrl: String? = nil,
        priceCents: Int = 0,
        currency: String = "USD",
        status: String = "published",
        enrollmentCount: Int? = nil,
        createdAt: String? = nil,
        updatedAt: String? = nil
    ) {
        self.id = id
        self.creatorId = creatorId
        self.creatorName = creatorName
        self.title = title
        self.slug = slug
        self.description = description
        self.coverImageUrl = coverImageUrl
        self.priceCents = priceCents
        self.currency = currency
        self.status = status
        self.enrollmentCount = enrollmentCount
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

extension Course: TableRecord, FetchableRecord, PersistableRecord {
    static let databaseTableName = "courses"

    enum Columns: String, ColumnExpression {
        case id
        case creatorId = "creator_id"
        case creatorName = "creator_name"
        case title
        case slug
        case description
        case coverImageUrl = "cover_image_url"
        case priceCents = "price_cents"
        case currency
        case status
        case enrollmentCount = "enrollment_count"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    enum CodingKeys: String, CodingKey {
        case id
        case creatorId = "creator_id"
        case creatorName = "creator_name"
        case title
        case slug
        case description
        case coverImageUrl = "cover_image_url"
        case priceCents = "price_cents"
        case currency
        case status
        case enrollmentCount = "enrollment_count"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

// MARK: - CoursePhase

struct CoursePhase: Codable, Identifiable, Equatable {
    var id: String
    var courseId: String
    var title: String
    var description: String?
    var sortOrder: Int

    enum CodingKeys: String, CodingKey {
        case id
        case courseId = "course_id"
        case title
        case description
        case sortOrder = "sort_order"
    }
}

// MARK: - PhaseModule

struct PhaseModule: Codable, Identifiable, Equatable {
    var id: String
    var phaseId: String
    var title: String
    var description: String?
    var sortOrder: Int
    var isPreview: Int // 0 or 1

    enum CodingKeys: String, CodingKey {
        case id
        case phaseId = "phase_id"
        case title
        case description
        case sortOrder = "sort_order"
        case isPreview = "is_preview"
    }
}

// MARK: - ContentBlock

struct ContentBlock: Codable, Identifiable, Equatable {
    var id: String
    var moduleId: String
    var blockType: String // "video", "text", "quiz", "image", "download", "code"
    var sortOrder: Int
    var content: String // JSON string

    enum CodingKeys: String, CodingKey {
        case id
        case moduleId = "module_id"
        case blockType = "block_type"
        case sortOrder = "sort_order"
        case content
    }
}

// MARK: - CourseEnrollment

struct CourseEnrollment: Codable, Identifiable, Equatable {
    var id: String
    var courseId: String
    var userId: String
    var pricePaidCents: Int
    var enrolledAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case courseId = "course_id"
        case userId = "user_id"
        case pricePaidCents = "price_paid_cents"
        case enrolledAt = "enrolled_at"
    }
}

extension CourseEnrollment: TableRecord, FetchableRecord, PersistableRecord {
    static let databaseTableName = "course_enrollments"

    enum Columns: String, ColumnExpression {
        case id
        case courseId = "course_id"
        case userId = "user_id"
        case pricePaidCents = "price_paid_cents"
        case enrolledAt = "enrolled_at"
    }
}

// MARK: - ModuleCompletion

struct ModuleCompletion: Codable, Identifiable, Equatable {
    var id: String
    var moduleId: String
    var userId: String
    var completedAt: String?
    var quizAnswers: String?

    enum CodingKeys: String, CodingKey {
        case id
        case moduleId = "module_id"
        case userId = "user_id"
        case completedAt = "completed_at"
        case quizAnswers = "quiz_answers"
    }
}

// MARK: - Full Course Tree (API response)

struct CourseWithContent: Codable {
    let course: Course
    let phases: [PhaseWithModules]
}

struct PhaseWithModules: Codable, Identifiable {
    var id: String { phase.id }
    let phase: CoursePhase
    let modules: [ModuleWithBlocks]
}

struct ModuleWithBlocks: Codable, Identifiable {
    var id: String { module.id }
    let module: PhaseModule
    let blocks: [ContentBlock]
}

struct CourseProgress: Codable {
    let totalModules: Int
    let completedModules: Int

    enum CodingKeys: String, CodingKey {
        case totalModules = "total_modules"
        case completedModules = "completed_modules"
    }
}
